File: src/minecraft/net/machinemuse/api/electricity/ElectricItemUtils.java
package net.machinemuse.api.electricity;

import java.util.ArrayList;
import java.util.List;

import net.machinemuse.powersuits.item.ItemComponent;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;

public class ElectricItemUtils {

	public static final String MAXIMUM_ENERGY = "Maximum Energy";
	public static final String CURRENT_ENERGY = "Current Energy";

	private static List<ElectricAdapter> electricItemsEquipped(EntityPlayer player) {
		List<ElectricAdapter> electrics = new ArrayList<ElectricAdapter>();
		for (int i = 0; i < player.inventory.getSizeInventory(); i++) {
			ItemStack stack = player.inventory.getStackInSlot(i);
			ElectricAdapter adapter = ElectricAdapter.wrap(stack);
			if (adapter != null) {
				electrics.add(adapter);
			}
		}
		return electrics;
	}

	public static double getPlayerEnergy(EntityPlayer player) {
		double avail = 0;
		for (ElectricAdapter adapter : electricItemsEquipped(player)) {
			avail += adapter.getCurrentEnergy();
		}
		return avail;
	}

	public static double getMaxEnergy(EntityPlayer player) {
		double avail = 0;
		for (ElectricAdapter adapter : electricItemsEquipped(player)) {
			avail += adapter.getMaxEnergy();
		}
		return avail;
	}

	public static void drainPlayerEnergy(EntityPlayer player, double drainAmount) {
		for (ElectricAdapter adapter : electricItemsEquipped(player)) {
			double drained = adapter.drainEnergy(drainAmount);
			drainAmount -= drained;
		}
	}

	public static void givePlayerEnergy(EntityPlayer player, double joulesToGive) {
		for (ElectricAdapter adapter : electricItemsEquipped(player)) {
			double given = adapter.giveEnergy(joulesToGive);
			joulesToGive -= given;
		}
	}

	public static double jouleValueOfComponent(ItemStack stackInCost) {
		if (stackInCost.getItem() instanceof ItemComponent) {
			if (stackInCost.getItemDamage() == ItemComponent.lvcapacitor.getItemDamage()) {
				return 20000 * stackInCost.stackSize;
			}
			if (stackInCost.getItemDamage() == ItemComponent.mvcapacitor.getItemDamage()) {
				return 100000 * stackInCost.stackSize;
			}
			if (stackInCost.getItemDamage() == ItemComponent.hvcapacitor.getItemDamage()) {
				return 750000 * stackInCost.stackSize;
			}
		}
		return 0;
	}

	public static double getPlayerHeat(EntityPlayer player) {
		return 0;
	}

	public static double getMaxHeat(EntityPlayer player) {
		return 1;
	}

}

File: src/minecraft/net/machinemuse/utils/ElectricItemUtils.scala
package net.machinemuse.utils

import net.machinemuse.api.electricity.ElectricAdapter
import net.machinemuse.powersuits.item.ItemComponent
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack

object ElectricItemUtils {
  def electricItemsEquipped(player: EntityPlayer): Seq[ElectricAdapter] = {
    val electrics = scala.collection.mutable.LinkedList
    for (i <- 0 until player.inventory.getSizeInventory) {
      val adapter: ElectricAdapter = ElectricAdapter.wrap(player.inventory.getStackInSlot(i))
      if (adapter != null) {
        electrics.add(adapter)
      }
    }
    electrics
  }

  def getPlayerEnergy(player: EntityPlayer): Double = {
    0 /: electricItemsEquipped(player) {
      case (avail, adapter) => avail + adapter.getCurrentEnergy
    }
  }

  def getMaxEnergy(player: EntityPlayer): Double = {
    0 /: electricItemsEquipped(player) {
      case (avail, adapter) => avail + adapter.getMaxEnergy
    }
  }

  def drainPlayerEnergy(player: EntityPlayer, drainAmount: Double) {
    drainAmount /: electricItemsEquipped(player) {
      case (drainleft, adapter) => drainleft - adapter.drainEnergy(drainleft)
    }
  }

  def givePlayerEnergy(player: EntityPlayer, joulesToGive: Double) {
    joulesToGive /: electricItemsEquipped(player) {
      case (joulesleft, adapter) => joulesleft - adapter.giveEnergy(joulesleft)
    }
  }

  def jouleValueOfComponent(stackInCost: ItemStack): Double = {
    stackInCost.getItem match {
      case i: ItemComponent => stackInCost.getItemDamage match {
        case ItemComponent.lvcapacitor.getItemDamage => 20000 * stackInCost.stackSize
        case ItemComponent.mvcapacitor.getItemDamage => 100000 * stackInCost.stackSize
        case ItemComponent.hvcapacitor.getItemDamage => 750000 * stackInCost.stackSize
        case _ => 0
      }
      case _ => 0
    }
  }

  final val MAXIMUM_ENERGY: String = "Maximum Energy"
  final val CURRENT_ENERGY: String = "Current Energy"
}